File: 0673-number-of-longest-increasing-subsequence/0673-number-of-longest-increasing-subsequence.java
class Solution {
    public int findNumberOfLIS(int[] nums) {
        int n = nums.length;
        if (n == 0) return 0;
        int[] lis = new int[n]; 
        int[] cnt = new int[n]; 
        for (int i = 0; i < n; i++) {
            lis[i] = 1;
            cnt[i] = 1;
        }
        int maxLength = 1, result = 0;
        for (int i = 1; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (nums[i] > nums[j]) {
                    if (lis[j] + 1 > lis[i]) {
                        lis[i] = lis[j] + 1;
                        cnt[i] = cnt[j]; 
                    } else if (lis[j] + 1 == lis[i]) {
                        cnt[i] += cnt[j]; 
                    }
                }
            }
            maxLength = Math.max(maxLength, lis[i]);
        }
        for (int i = 0; i < n; i++) {
            if (lis[i] == maxLength) {
                result += cnt[i];
            }
        }
        return result;
    }
}

File: 0673-number-of-longest-increasing-subsequence/0673-number-of-longest-increasing-subsequence.cpp
class Solution {
public:
    int findNumberOfLIS(vector<int>& arr) {
        int n = arr.size();
        vector<int> dp(n, 1), cnt(n, 1);
        int maxLen = 1;
        for (int i = 0; i < n; i++) {
            for (int prev = 0; prev < i; prev++) {
                if (arr[i] > arr[prev]) {
                    if (dp[prev] + 1 > dp[i]) {
                        dp[i] = dp[prev] + 1;
                        cnt[i] = cnt[prev]; 
                    } else if (dp[prev] + 1 == dp[i]) {
                        cnt[i] += cnt[prev]; 
                    }
                }
            }
            maxLen = max(maxLen, dp[i]);
        }
        int count = 0;
        for (int i = 0; i < n; i++) {
            if (dp[i] == maxLen) {
                count += cnt[i];
            }
        }
        return count;
    }
};
